### PR TITLE
Optimize the fbits2i Evaluator

### PR DESCRIPTION
**Pull request by dchopra001**
The fbits2i evaluator in the z codegen is suboptimal and hence causes a performance regression as compared to the x codegen. This commit captures the missing corner cases to close that gap, and includes some extra cleanup to simplify the evaluator.

### CODE REVIEW

**Comment by dchopra001** [comment]
Marked this as a WIP PR as there's still some extra work (and cleanup) required here. Once this is ready for review I'll remove the WIP tag.

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 1]
```diff
@@ -1204,86 +1204,119 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
+          targetReg = genericLoadHelper<32, 32, MemReg>(node, cg, tempMR2, NULL, false, true);
+          // TODO: I think we can remove this as it's handled automatically for us. 
+          tempMR2->stopUsingMemRefRegister(cg); 
+          }
+       }
+    
+    //TODO: for an fbits2i node, this should always be set. However in my unit test these instructions
+    // never get generated. I believe the same problem occurs on X. We need to investigate why. 
+    // This should always be set for a java program.
+    if (node->normalizeNanValues())
```
Comment (dchopra001) [inline compiler/z/codegen/FPTreeEvaluator.cpp]: @fjeremic The x86 evaluator has a similar guard. As can be seen by this comment: 
https://github.com/eclipse/openj9/issues/890#issuecomment-367805534, we never generated the normalization instructions and we should have (this is for a java program). 

I checked inside openj9, and we set this normalization parameter to true when we inline the function call here: https://github.com/eclipse/openj9/blob/b2b62046eb4c221ad1773457bae20da74f08ec09/runtime/compiler/env/VMJ9.cpp#L7813-L7818

So it should always be true for a java program.

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 2]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    TR::Register * targetReg;
    TR::Register * sourceReg;
    TR::Instruction * cursor = NULL;
-   TR::RegisterDependencyConditions * deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
-   //Disable FPE Version For Now Due To Performance Reasons
-   bool disabled = true;
-   sourceReg = cg->evaluate(node->getFirstChild());
 
+   TR::MemoryReference  *tempmemref;
    TR::SymbolReference * f2iSR = cg->allocateLocalTemp(TR::Int32);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: 122216ccd77679f8e9bf4ad959dc779cf308c6b1

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 3]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
+   bool disableFPInstructions = true;
+
+   // if child is a load with refCount == 1, then just use a regular load
+   // to load as an integer 
+   if (node->getFirstChild()->getRegister() == NULL &&
+          node->getFirstChild()->getOpCode().isLoadVar() &&
+          (node->getFirstChild()->getReferenceCount() == 1))
       {
-      deps->addPostCondition(litBase, TR::RealRegister::AssignAny);
+      tempmemref = generateS390MemoryReference(node->getFirstChild(), cg);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: 122216ccd77679f8e9bf4ad959dc779cf308c6b1

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 4]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
-   TR::MemoryReference * NaN              = generateS390MemoryReference((int32_t)0x7fc00000, TR::Int32, cg, litBase);
-   if (litBase)
+   //Disable FPE Version For Now Due To Performance Reasons
+   bool disableFPInstructions = true;
+
+   // if child is a load with refCount == 1, then just use a regular load
+   // to load as an integer 
+   if (node->getFirstChild()->getRegister() == NULL &&
+          node->getFirstChild()->getOpCode().isLoadVar() &&
+          (node->getFirstChild()->getReferenceCount() == 1))
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: 021b0977bb94669bd4fb30e1c7c2e7b27d9a0186

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 5]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
-      }
-   cleansedNumber->setEndInternalControlFlow();
-   generateRXInstruction(cg, TR::InstOpCode::TCEB, node, sourceReg, (uint32_t) 0x03f);
-   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, cleansedNumber);
-   cursor->setStartInternalControlFlow();
-   TR::MemoryReference * positiveInfinity = generateS390MemoryReference((int32_t)0x7f800000, TR::Int32, cg, litBase);
-   TR::MemoryReference * negativeInfinity = generateS390MemoryReference((int32_t)0xff800000, TR::Int32, cg, litBase);
-   TR::MemoryReference * NaN              = generateS390MemoryReference((int32_t)0x7fc00000, TR::Int32, cg, litBase);
-   if (litBase)
+   //Disable FPE Version For Now Due To Performance Reasons
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: 74625a72

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 6]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    TR::Register * targetReg;
    TR::Register * sourceReg;
    TR::Instruction * cursor = NULL;
-   TR::RegisterDependencyConditions * deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
-   //Disable FPE Version For Now Due To Performance Reasons
-   bool disabled = true;
-   sourceReg = cg->evaluate(node->getFirstChild());
 
+   TR::MemoryReference  *tempmemref;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: 74625a72

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 7]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
+    if (litBase)
+       {
+       deps->addPostCondition(litBase, TR::RealRegister::AssignAny);
+       }
+    deps->addPostCondition(targetReg, TR::RealRegister::AssignAny);
+
+    generateRXInstruction(cg, TR::InstOpCode::TCEB, node, targetReg, (uint32_t) 0x03f);
+
+    cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BZ, node, cleansedNumber);
+    cursor->setStartInternalControlFlow();
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: Hmpf. That's an interesting observation. It seems we didn't even have a label instruction for the start of ICF. This instruction was meant to accomplish that. I've made the change to make this more consistent with the other codegens, along with some minor cleanup here: ff8c0b05

File: compiler/z/codegen/FPTreeEvaluator.cpp [hunk 8]
```diff
@@ -1204,86 +1204,113 @@ OMR::Z::TreeEvaluator::fbits2iEvaluator(TR::Node * node, TR::CodeGenerator * cg)
...
+         generateRSInstruction(cg, TR::InstOpCode::SRLG, node, targetReg, targetReg, 32);
+         }
+      else
+         {
+         // It seems the FP instruction performance is sub-optimal (i.e. LGDR), and we're better
+         // off storing to memory and then loading into a GPR. 
+         TR::MemoryReference * tempMR1 = generateS390MemoryReference(node, f2iSR, cg);
+         generateRXInstruction(cg, TR::InstOpCode::STE, node, sourceReg, tempMR1);
+
+         targetReg = genericLoadHelper<32, 32, MemReg>(node, cg, tempMR1, NULL, false, true);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: I don't have any concrete data right now to back this up, however in a quick run with a simple java benchmark, I did observe significantly slower performance when enabling the FP instructions path. I'll collect more data and report back here with official results.

Also, as a side node, this pattern of disabling FP instructions is seen in other similar evaluators. Here's a few examples:

https://github.com/eclipse/omr/blob/6ecb3f3483722386111ec8049bd8a90e5c05f6d0/compiler/z/codegen/FPTreeEvaluator.cpp#L1169-L1184

https://github.com/eclipse/omr/blob/6ecb3f3483722386111ec8049bd8a90e5c05f6d0/compiler/z/codegen/FPTreeEvaluator.cpp#L1294-L1311

https://github.com/eclipse/omr/blob/6ecb3f3483722386111ec8049bd8a90e5c05f6d0/compiler/z/codegen/FPTreeEvaluator.cpp#L1358-L1363

[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: So I ran the benchmark described here: https://github.com/eclipse/openj9/issues/890#issuecomment-407955069

On the following OpenJ9 version:
```
OpenJ9   - 4e11189
OMR      - e09bd01
```

This did *not* include my changes in the PR here. This is using the implementation of the `fbits2i` evaluator as it currently exists in the `eclipse/omr` repository. 

My results are below:
screen shot 2018-08-08 at 10 31 43 am

Based on the above, we are ~1.1% faster when the floating point instructions path is disabled.

[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: Interesting, thanks for the data backing up your claims.

File: compiler/z/codegen/FPTreeEvaluator.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: @dchopra001 not sure if you understood my previous message. We currently don't set ICF flags on labels in the Z codegen. ff8cb0 seems to do that. Can we set the flags on the instructions as is above in the review on line 1281? We did it here for the start of the ICF but not for the end.

[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: I've opened up #2833 to address this in the future.

[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: Hmm, this is strange. I don't see any consistency here. I see lots of examples in the Z codegen where we prefer setting the ICF Flag on the label. Here's a few examples:
https://github.com/eclipse/omr/blob/ab706f4268160c4a46bb701c0558e6c9055afc97/compiler/z/codegen/BinaryEvaluator.cpp#L772-L781

https://github.com/eclipse/omr/blob/ab706f4268160c4a46bb701c0558e6c9055afc97/compiler/z/codegen/BinaryEvaluator.cpp#L3988-L4002

https://github.com/eclipse/omr/blob/ab706f4268160c4a46bb701c0558e6c9055afc97/compiler/z/codegen/OMRTreeEvaluator.cpp#L7719

https://github.com/eclipse/omr/blob/ab706f4268160c4a46bb701c0558e6c9055afc97/compiler/z/codegen/OMRTreeEvaluator.cpp#L12265

At least inside `OMRTreeEvaluator.cpp`, it looks like we prefer setting the `ending ICF` flag on the label as opposed to setting it on the instruction (the start control flow flag is a little more balanced). Perhaps the cleanup required as part of #2833 is to change the locations where we set the flag on the instruction only. Do you still prefer that we switch to setting it on the instruction in this case?

[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/FPTreeEvaluator.cpp]: > Do you still prefer that we switch to setting it on the instruction in this case?

No, I did not spot all these label ICF setters in my original grep. Looks like we are consistently inconsistent here. It should be fine to leave it as is until #2833 is addressed. Good catch! Thanks for spotting that.

**Comment by fjeremic** [comment]
@dchopra001 could you squash the commits and I'll kick off a testing build.

**Comment by dchopra001** [comment]
@fjeremic Done!

**Comment by fjeremic** [comment]
Merging before builds are finished as this is a Z only change and we don't have Z testing at AppVeyor or TravisCI. The OMR Z builds have passed successfully.

**Pull request merged**